### PR TITLE
Handle non-existent posts

### PR DESCRIPTION
**Pull request by maartenmachiels**
Add a check to see if the blogpost exists, and if not, throw a 404 error.

### CODE REVIEW

File: components/Post.php [hunk 1]
```diff
@@ -103,6 +103,14 @@ protected function loadPost()
...
             });
         }
+        
+        /*
+         * Throw a 404 error when a blogpost does not exist
+         */
+        if (!$post || !$post->exists) {
+            $this->setStatusCode(404);
+            return $this->controller->run('404');
+        }
```
Comment (bennothommo) [inline components/Post.php]: @maartenmachiels The block of code above should be run immediately after trying to retrieve the post. On this line here (https://github.com/rainlab/blog-plugin/blob/0b4ef4609fd06f8b83d9beca8149d2cd32cc5850/components/Post.php#L96), you could change the call to `firstOrFail()` and wrap the line in a `try..catch` block. If a `Illuminate\Database\Eloquent\ModelNotFoundException` exception is caught, you can then run your 404 code.

**Comment by maartenmachiels** [comment]
Hi @bennothommo thank you for your suggestions. I've changed the code here https://github.com/maartenmachiels/blog-plugin/blob/patch-1/components/Post.php 

I hope I've done everything by the standards. If you require anything else, please let me know and I'll get right on it.

File: components/Post.php [hunk 2]
```diff
@@ -93,7 +93,24 @@ protected function loadPost()
             $post = $post->isPublished();
         }
 
-        $post = $post->first();
+        try {
+            $post = $post->firstOrFail();
```
Comment (bennothommo) [inline components/Post.php]: @maartenmachiels Remove everything below this line in the `try` block, as the same code is specified underneath the `catch` block.

**Comment by bennothommo** [comment]
Thanks @maartenmachiels. Just one more change needed, and then I will test it :)

**Comment by maartenmachiels** [comment]
Hi @bennothommo , I'm sorry for that. Looks like my head isn't screwed on right this morning. I've edited the code. Anything else?

**Comment by bennothommo** [comment]
@maartenmachiels My apologies - I've lead you down the wrong path. :confused: I've just realised that the October Query builder does not have a `firstOrFail()` method. I will probably need to rectify this at some point, but for now, would you mind reverting your PR back to the state it was at 0b4ef4609fd06f8b83d9beca8149d2cd32cc5850. I'll test it from there.

**Comment by bennothommo** [comment]
@maartenmachiels Please disregard last comment - it does have that method. Sorry for the mess-around. I'll begin testing now.

**Comment by bennothommo** [comment]
It appears my head isn't screwed on right either. :P

**Comment by bennothommo** [comment]
@maartenmachiels Tested and it works well. I've made a slight adjustment only so that a 404 is thrown only if `ModelNotFoundException` is thrown, so that other exceptions are not caught but are instead handled in the default October way (show error screen when debug mode is on, or show generic server error message if debug mode is off).

Thank you for your contribution! :)

**Comment by maartenmachiels** [comment]
Thank you so much @bennothommo for your work and testing. And thanks to the OP for their original pull request on which this is solution is based.

**Comment by alxy** [comment]
@bennothommo How it this supposed to work? I tried to apply this very same code to my custom component, but it doesnt work, as the ``onRun()`` function would need to return the 404 response, and not the ``loadPost()`` function...

**Pull request merged**